### ISSUE TITLE
multi built-ins are not assignable to Callable

### ISSUE THREAD

**Issue by 0racle** [rakudo/rakudo#1566]
## The Problem

When assigning a built-in function (eg. `max`) to a Callable, Rakudo complains that it "expected Callable but got Sub+{is-pure}"

## Expected Behavior

```perl-6
my &func = &max;
say func([3, 5, 2]);    # OUTPUT: 5

# OR

sub foo(&func, |c) {
    return func(|c);
}

say foo(&max, [3, 5, 2]);    # OUTPUT: 5
```

## Actual Behavior

Since `max` and other built-ins were turned into multi's (eg. [55bc053](https://github.com/rakudo/rakudo/commit/55bc053c56bd54717fe39c627a539373aba83013)) trying to do either of the above yields the following runtime error

```
Type check failed in assignment to &func;
expected Callable but got Sub+{is-pure} (Sub+{is-pure}.new)
```

**Comment by 0racle** [comment]
I created my own pure multi sub in a module and exported it and it did not suffer from the same issue. Despite identifying as a `(Sub+{is-pure})` I could still assign to Callable vars just fine.

```perl6
unit module Mux;

proto sub mux(|) is pure is export {*} 
multi sub mux(+args, :&by!) { args.max(&by) }
multi sub mux(+args)        { args.max      } 
```
... elsewhere

```perl6
use lib 'lib';
use Mux;

say &mux.WHAT;    # OUTPUT: (Sub+{is-pure})
my &func = &mux;
say func([3, 5, 2]);    # OUTPUT: 5
```

**Comment by AlexDaniel** [comment]
Blocker label added to make sure we do something about this before the release.

**Comment by zoffixznet** [comment]
The title of the Issue is inaccurate, you *can* assign core multies:

```
<Zoffix___> m: my &foo = &say
<camelia> rakudo-moar 2a8b17ae9: ( no output )
<Zoffix___> m: my &foo = &say; say +&say.candidates
<camelia> rakudo-moar 2a8b17ae9: OUTPUT: «5␤»
```

The issue looks to be the same as [RT#128905](https://rt.perl.org/Ticket/Display.html?id=128905#ticket-history). Seems https://github.com/rakudo/rakudo/commit/55bc053c56bd54717fe39c627a539373aba83013 made `&max`/`&min`/`&minmax` join the ranks of affected routines.

**Comment by zoffixznet** [comment]
Found that composing affected types fixes the issue:

```
<Zoffix_> m: &min.^compose; my &f = &min;
<camelia> rakudo-moar 838782b7d: ( no output )
```

and Perl6::Metamodel::Mixins.mixin has this comment in it:

```
        # XXX Workaround for mixing in to non-composed types; when this takes
        # place (a bunch during CORE.setting) the mixin is missing bits. This
        # has long been a problem, and needs a real solution 
```

So I guess the work-around doesn't work-around perfectly.

**Comment by lizmat** [comment]
FWIW, I believe the underlying issue (aka golf) is:

    use nqp; say nqp::istype(&max,Callable)   # 0 instead of 1

**Comment by lizmat** [comment]
I tried fixing this by moving `Routine.pm` to before `traits.pm` so that Routine would be fully composed.  But this turned out to be a rabbit-hole that will take some real re-organizing the setting for that to work, as it would also need `Attribute.pm` before `Routine.pm` and that is really a can of worms.

**Comment by zoffixznet** [comment]
Started on a promising endeavor of moving all subs into a place past all Callable types in the setting: https://github.com/rakudo/rakudo/commit/3506395576

...but we have traits on a bunch of methods too and dealing with those is trickier. I started adding `.^compose` through methods we used traits on, BUT, the types that INHERITED those methods would need that done too. So now I can only think of going through ALL methods of all types declared by the `src/core/traited-routines.pm6` point in the setting and `.^compose`ing all the ones that `!~~ Callable` False.

I've no real idea of what `.^compose` does. Is doing that stuff for methods that fail Callable typecheck  by that point too much?

**Comment by zoffixznet** [comment]
We didn't have the time to fix this before the release. Since conditions that cause the bug is known (mixing in a role into `Callable`s early enough in the setting before `Callable` is composed),  what I done is looked through [commits since last release](https://github.com/rakudo/rakudo/compare/2018.02...4513c279d4) and found all of the routines that acquired this bug *since last release*. These appear to be only `&min`, `&max`, and `&minmax`.

For those three, I [added a workaround](https://github.com/rakudo/rakudo/commit/4513c279d457fc4c27c3f3db4ba5f2e71e349b0d) and [tests covering them](https://github.com/perl6/roast/commit/154a22c9635d341149e07d863b5f049134a5ba65), so this ticket is no longer a release blocker, but the bug still needs to be fixed and more thoroughly tested.

**Comment by zoffixznet** [comment]
Heh. Well, time flew by, eh? Reading [the logs](https://irclog.perlgeek.de/perl6/2018-04-17#i_16058000), `&lc` and likely [a bunch more](https://github.com/rakudo/rakudo/commit/cdb45fa5a65f06b543db7939fc51af7b598e0bfb) are affected this release.

**Comment by 0racle** [comment]
Have found some more fallout from the multi-fication of subs.

```
> say dir.sort(&lc)
Cannot resolve caller lc(IO::Path, IO::Path); none of these signatures match:
    (Cool $s)
```

Additional context
```
bisectable • perlawhirl, bisect log: https://gist.github.com/f1dd78f91bf09798c9b9f5eb4cce8b3b
bisectable • perlawhirl, (2018-02-21) https://github.com/rakudo/rakudo/commit/cdb45fa5a65f06b543db7939fc51af7b598e0bfb
```

It seems the proto sig `(|)` causes 2 `IO::Path` values to be passed, through despite the fact that `IO::Path ~~ Cool`.

Current workaround is to force a single value, eg. `sort(*.lc)`, `sort({ lc($_ })`.

A possible quick-fix might be to introduce a another multi, ie.
```perl6
multi sub lc(Cool $a, Cool $b) {
    $a.lc cmp $b.lc
}
```

**Comment by zoffixznet** [comment]
Ah, that's an entirely different issue from OP. I filed it here: https://github.com/rakudo/rakudo/issues/1739

**Comment by Kaiepi** [comment]
Fixed with 7f2ae26cdfda0a598a7b3c9fa513000553490640.
